courseTasks/executor/executor.go

```diff
@@ -0,0 +1,103 @@
+package executor
+
+import (
+	"fmt"
+	"johnSamilin/golang-hangman/courseTasks/analyzer"
+	"johnSamilin/golang-hangman/courseTasks/executor/constants"
+	"johnSamilin/golang-hangman/courseTasks/utils"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+func startExecution(configId int, workers chan Worker, commands []string, exit chan bool, wg *sync.WaitGroup) analyzer.Stats {
+	var tasksPool = make(chan string, len(commands))
+	collector := analyzer.New()
+	for _, cmd := range commands {
+		tasksPool <- cmd
+		collector.AddLine(cmd)
+	}
+	for {
+		freeWorker := <-workers
+		if len(tasksPool) > 0 {
+			shouldStop := processCommand(freeWorker.Id, freeWorker.Lock, configId, <-tasksPool, exit)
+			if !shouldStop {
+				workers <- freeWorker
+			}
+			if shouldStop {
+				break
+			}
+		}
+		if len(tasksPool) == 0 {
+			close(tasksPool)
+			wg.Done()
+			break
+		}
+	}
+	return collector.GetStats()
+}
+
+func processCommand(idx int, lock *sync.Mutex, configId int, cmd string, exit chan bool) (shouldStop bool) {
+	shouldStop = false
+	var args = strings.SplitN(cmd, " ", 3)
+	lock.Lock()
+	commonInfo := "Worker #" + strconv.Itoa(idx) + " (task from cmd pool #" + strconv.Itoa(configId) + ")"
+	switch args[0] {
+	case constants.Time:
+		fmt.Println(commonInfo, "Time:", time.Now())
+		lock.Unlock()
```

Comment:
I see lock.Unlock at the end of each case. Would it be easier to put that into `defer` of some kind?

Reply:
defer won't work because TimeWait instruction starts it's own goroutine and unlocks mutex there. Having time.Sleep instead of goroutine also doesn't fit because it stops all other workers

Reply:
I saw that particular case but you have to find a way to make that particular worker sleep, this is the idea of TimeWait task.

I think the issue should disappear once you refactor tasks into separate entities (or it will be transparent on what should be changed in the design to support the main logic)